### PR TITLE
GH-37437: [C++] Fix MakeArrayOfNull for list array with large string values type

### LINKED ISSUE #37437: [C++] "utf8_split_whitespace" kernel returned offset buffers are too small for large string in case of empty array
### Describe the bug, including details regarding any error messages, version, and platform.

Consider the following call to `utf8_split_whitespace`
```python
import pyarrow as pa
import pyarrow.compute as pc

result = pc.utf8_split_whitespace(
    pa.array([], type=pa.large_string()),
)
print(result.type)
print(result.buffers())
```

The returned type is `list<item: large_string>`, but the returned offset buffer has `size=4`, not `size=8`. I expect to see `size=8`, pertaining to a length-one array of `int64` (a single zero)

For a non-empty array, the size is correct.

We've encountered this in Awkward Array (https://github.com/scikit-hep/awkward/pull/2679#issuecomment-1697469490).

Thanks!

### Component(s)

Python

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
### Rationale for this change

`MakeArrayOfNull` for list type was assuming that the values child field didn't need to be considered, but those values could also require a minimum buffer size (eg for offsets) and which could be of greater size than the list offsets if those are int32 offsets.

### Are these changes tested?

Yes

* Closes: #37437

### CODE REVIEW

File: cpp/src/arrow/array/validate.cc [hunk 1]
```diff
@@ -713,8 +713,10 @@ struct ValidateArrayImpl {
...
 
     // An empty list array can have 0 offsets
-    const auto required_offsets = (data.length > 0) ? data.length + data.offset + 1 : 0;
     const auto offsets_byte_size = data.buffers[1]->size();
+    const auto required_offsets = ((data.length > 0) || (offsets_byte_size > 0))
+                                      ? data.length + data.offset + 1
+                                      : 0;
     if (offsets_byte_size / static_cast<int32_t>(sizeof(offset_type)) <
         required_offsets) {
       return Status::Invalid("Offsets buffer size (bytes): ", offsets_byte_size,
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/array/validate.cc:722]: > @pitrou do you know if this "empty list array can have 0 offsets" also applies to other types that have offsets, i.e. string types? (this was added in #6230)

I don't think so. In any case, the allowance for list types is IMO historical.

> At the moment the offsets buffer size is not checked if the length of the array is 0, but I changed this to actually check it if the offset buffer has _some_ length (so assuming that _if_ there is an offsets buffer with a non-zero length, it should have a correct length, also if the array itself has length 0)

That sounds like a good approach.

File: cpp/src/arrow/array/array_test.cc [hunk 2]
```diff
@@ -388,6 +388,7 @@ static std::vector<std::shared_ptr<DataType>> TestArrayUtilitiesAgainstTheseType
       large_utf8(),
       list(utf8()),
       list(int64()),  // NOTE: Regression case for ARROW-9071/MakeArrayOfNull
+      list(large_utf8()),
```
Comment (pitrou) [inline cpp/src/arrow/array/array_test.cc:391]: Can we also add some nested list types into the mix?

Comment (jorisvandenbossche) [reply cpp/src/arrow/array/array_test.cc:391]: I added a few more nested combinations

**Review by pitrou — APPROVED** [review]
+1, thank you!

**Pull request merged**